Use of host function 'deallocate' in a host/device function error in c++20 code built for AMD GPU

We have an error building a Kokkos-based app for AMD MI300A GPUs. We've reduced the code to a few lines while still getting the same error message. The following code produces an error when using the AMD gfx942 target.

```c++
#include <string> 
class LabeledObject 
{ std::string label; public: LabeledObject(); virtual ~LabeledObject(); } 
; 
class Describable : virtual public LabeledObject 
{ public: virtual ~Describable(); virtual std::string describe() = 0; } 
; 
template <class Scalar> 
class ScalarProdBase : virtual public Describable 
{ 
public: 
std::string describe() 
{ return "foo"; } 
protected: 
virtual bool bar() = 0; 
}; 
template class ScalarProdBase<double>;
```

I'm using the following command to build:

`clang++ --std=c++20 -xhip -o repro.o -c repro.cpp --offload-arch=gfx942`

Notably, the error does not happen when using --std=c++17, but the code need c++20 for other parts.

This is the error output I see:

```
In file included from <built-in>:1:
In file included from /llvm-dir/lib/clang/23/include/__clang_hip_runtime_wrapper.h:151:
In file included from /llvm-dir/lib/clang/23/include/cuda_wrappers/complex:35:
In file included from /opt/rh/gcc-toolset-13/root/usr/lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/stdexcept:39:
In file included from /opt/rh/gcc-toolset-13/root/usr/lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/string:43:
/opt/rh/gcc-toolset-13/root/usr/lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/allocator.h:210:25: error: 
      reference to __host__ function 'deallocate' in __host__ __device__
      function
  210 |         __allocator_base<_Tp>::deallocate(__p, __n);
      |
```

--- a/clang/lib/Sema/SemaCUDA.cpp
+++ b/clang/lib/Sema/SemaCUDA.cpp
@@ -496,7 +496,9 @@ bool SemaCUDA::inferTargetForImplicitSpecialMember(CXXRecordDecl *ClassDecl,
 
   // Look for special members in base classes that should be invoked from here.
   // Infer the target of this member base on the ones it should call.
-  // Skip direct and indirect virtual bases for abstract classes.
+  // Skip direct and indirect virtual bases for abstract classes, except for
+  // destructors — the complete destructor variant destroys virtual bases
+  // regardless of whether the class is abstract.
   llvm::SmallVector<const CXXBaseSpecifier *, 16> Bases;
   for (const auto &B : ClassDecl->bases()) {
     if (!B.isVirtual()) {
@@ -504,9 +506,8 @@ bool SemaCUDA::inferTargetForImplicitSpecialMember(CXXRecordDecl *ClassDecl,
     }
   }
 
-  if (!ClassDecl->isAbstract()) {
+  if (!ClassDecl->isAbstract() || CSM == CXXSpecialMemberKind::Destructor)
     llvm::append_range(Bases, llvm::make_pointer_range(ClassDecl->vbases()));
-  }
 
   for (const auto *B : Bases) {
     auto *BaseClassDecl = B->getType()->getAsCXXRecordDecl();

--- a/clang/test/SemaCUDA/dtor-constexpr-virtual-base.cu
+++ b/clang/test/SemaCUDA/dtor-constexpr-virtual-base.cu
@@ -1,0 +1,33 @@
+// RUN: %clang_cc1 %s -std=c++20 -fsyntax-only -verify=host
+// RUN: %clang_cc1 %s -std=c++20 -fcuda-is-device -fsyntax-only -verify=dev
+
+// host-no-diagnostics
+// dev-no-diagnostics
+
+#include "Inputs/cuda.h"
+
+// The implicit destructor of an abstract class with virtual bases should
+// consider those virtual bases during CUDA target inference, since the
+// complete destructor variant destroys them.
+void host_only();
+
+constexpr void wraps_host() {
+  if (!__builtin_is_constant_evaluated())
+    host_only();
+}
+
+struct HasDtor {
+  ~HasDtor() { wraps_host(); }
+};
+
+struct Base {
+  HasDtor m;
+  virtual ~Base();
+};
+
+template <class T>
+struct Derived : virtual public Base {
+  virtual void foo() = 0;
+};
+
+template class Derived<int>;

